puzzle-utils/words/dropquote.py

```diff
@@ -0,0 +1,193 @@
+        
+        column += 1
+        if column >= len(grid[row]):
+            row += 1
+            column = 0
+    
+    if current_word.find(' ') != -1:
+        # Overlapped columns could produce a superset of the actual possible words
+        snapshot.words[word_index] = DropQuoteWords(current_letters[:], dropquote_search_words(current_word, current_letters))
+
+    return snapshot
+
+def dropquote_search_words(word, letters):
+    pattern = '^'
+    for index in range(len(word)):
+        if word[index] == ' ':
+            pattern += f'[{letters[index]}]'
+        else:
+            pattern += word[index]
+    pattern += '$'
+    return search(pattern)
+
+def dropquote_apply_word(grid, letters, word_index, word):
+    """
+    Applies a word at the word index to the grid and removes the consumed letters.
+    >>> grid = ['    -', '  -  ', '     ']
+    >>> letters = ['gor', 'afo', 'mn', 'ete', 'hs']
+    >>> dropquote_apply_word(grid, letters, 2, 'thrones')
+    >>> print(grid)
+    ['    -', '  -th', 'rones']
+    >>> print(letters)
+    ['go', 'af', 'm', 'e', '']
+    """
+    row, column = dropquote_locate_word(grid, letters, word_index)
+    dropquote_apply_word_validate(grid, letters, word, row, column)
+    dropquote_apply_word_at_location(grid, letters, word, row, column)
+
+def dropquote_locate_word(grid, letters, word_index):
+    row = 0
+    column = 0
+    current_word_index = 0
+    new_word = True
+    while row < len(grid) and column < len(grid[0]):
+        if grid[row][column] == '-' and new_word == False:
+            current_word_index += 1
+            new_word = True
+        elif grid[row][column] != '-' and new_word == True:
+            new_word = False
+            if current_word_index == word_index:
+                break
+
+        column += 1
+        if column >= len(grid[row]):
+            row += 1
+            column = 0
+    
+    if row >= len(grid) or column > len(grid[0]):
+        raise ValueError('The word index is greater than the number of words.')
+    
+    return row, column
+
+def dropquote_apply_word_validate(grid, letters, word, row, column):
+    current_letters = letters[:]
+    for letter in word:
+        if letter == '-':
+            raise ValueError('The provided word was longer than the word in the grid.')
+
+        if letter != grid[row][column] and grid[row][column] != ' ':
+            raise ValueError('The word conflicts with existing letters in grid.')
+
+        if current_letters[column].find(letter) == -1:
+            raise ValueError(f'There were not enough \'{letter}\' found in column {column}.')
+
+        current_letters[column] = current_letters[column].replace(letter, '', 1)
+        column += 1
+        if column >= len(grid[row]):
+            row += 1
+            column = 0
+
+def dropquote_apply_word_at_location(grid, letters, word, row, column):
+    for letter in word:
+        grid[row] = grid[row][:column] + letter + grid[row][column + 1:]
```

Comment:
Would it be easier to work with lists of characters so your data structures are mutable? Here you could just assign/insert the thing you want. The input can still be strings, but you can list() it at the entry point for dropquote